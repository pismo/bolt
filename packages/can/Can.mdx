---
'name': Can
---

import {Fragment} from 'react'
import { Playground } from 'docz'
import {Can, StaticCan} from './Can'

# Can

O componente **Can** fornece uma forma para validar permisões do usuário. Pode ser utilizado como componente ou como um método javascript.

## StaticCan

**StaticCan** é um utilitario que possui a variável *ROLES* estática.
Pode ser usada para definir as permissões globais da aplicação, os componentes Can utilizam essa variável por padrão

```
import { StaticCan } from '@pismo/bolt-can'

StaticCan.ROLES = ["batata", "pimentão"]
```

## Can

O componente **Can** pode ser usado para exibir (ou não) um outro componente de acordo com as permissões do usuário

### Propriedades

| **prop** | **type**                                         | **required** | **desc**                                              |
| -------- | ------------------------------------------------ | :----------: | ----------------------------------------------------- |
| userRoles | string[]                                        |     true     | são as permissões do usuário, normalmente obtidas através de um login   |
| children  | React.RactNode                                  |     true     | os componentes que serão exibidos caso as permissões sejam válidas      |
| roles     | string[] = StaticCan.ROLES                      |     false    | as permissões necessárias para a validação atual, por default utiliza a variável estática *StaticCan.ROLES*  |
| fallback  | () => React.ReactNode | React.FC | null         |    false     | componente que será exibido caso as permissões não sejam válidas |

```
import { StaticCan, Can } from '@pismo/bolt-can'
```

<Playground>
    {
        () => {
            StaticCan.ROLES = ["batata", "cenoura"]
            return (
                <Fragment>
                    <Can userRoles={["batata"]} fallback={() => <div>não tenho permissão</div>}>
                        tenho permissão => batata
                    </Can>
                    <Can userRoles={["tomate"]} fallback={() => <div>não tenho permissão => tomate</div>}>
                        tenho permissão => batata
                    </Can>
                    <Can userRoles={["banana"]} roles={["banana", "maçã"]} fallback={() => <div>não tenho permissão</div>}>
                        tenho permissão => banana
                    </Can>
                </Fragment>
            )
        }
    }
</Playground>

## toCan

Método que pode ser utilizado para validar permissões, retorna verdadeiro caso as permissões sejam válidas ou false se não

### Propriedades

| **prop** | **type**                                         | **required** | **desc**                                              |
| -------- | ------------------------------------------------ | :----------: | ----------------------------------------------------- |
| userRoles | string[]                                        |     true     | são as permissões do usuário, normalmente obtidas através de um login   |
| roles     | string[] = StaticCan.ROLES                      |     false    | as permissões necessárias para a validação atual, por default utiliza a variável estática *StaticCan.ROLES*  |

```
import { StaticCan, toCan } from '@pismo/bolt-can'

//toCan(userRoles: string[], roles: string[] = StaticCan.ROLES) => boolean

StaticCan.ROLES = ["batata", "cenoura"]

toCan(["batata"]) // true
toCan(["tomate"]) // false
toCan(["tomate", "cenoura"]) // true
toCan(["banana"], ["maçã", "banana"]) // true
```