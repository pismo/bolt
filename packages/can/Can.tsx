import * as React from 'react'

const { Fragment } = React

export class StaticCan {
  static ROLES: string[] = []
}
export interface CanProps {
  userRoles: string[]
  children: React.ReactNode
  roles?: string[]
  fallback?: () => React.ReactNode | React.FC | null
}

export const toCan = (
  userRoles: string[],
  roles: string[] = StaticCan.ROLES
): boolean => userRoles.some(val => roles.indexOf(val) !== -1)

export const Can: React.FC<CanProps> = ({
  roles = StaticCan.ROLES,
  userRoles,
  children,
  fallback
}: CanProps) => (
  <Fragment>
    {toCan(userRoles, roles) ? children : fallback && fallback()}
  </Fragment>
)
