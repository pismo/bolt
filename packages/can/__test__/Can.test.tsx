import { cleanup, render, } from '@testing-library/react'
import 'jest-dom/extend-expect'
import * as React from 'react'

import { Can, toCan } from '../Can'

afterEach(cleanup)

describe('@pismo/bolt-can', () => {
  test('if the user has the permissions should display the components', () => {
    const {container} = render(<Can userRoles={["tomate"]} roles={["tomate"]}>permitido</Can>)

    expect(toCan(["tomate"], ["tomate"])).toBe(true)    
    expect(container).toHaveTextContent('permitido')
    expect(container).toMatchSnapshot()
  })

  test('if the user does not have the permissions should display the components', () => {
    const {container} = render(<Can userRoles={["tomate"]} roles={["cebola"]}>permitido</Can>)

    expect(toCan(["banana"], ["tomate"])).toBe(false)
    expect(container).not.toHaveTextContent('permitido')
  })
})
