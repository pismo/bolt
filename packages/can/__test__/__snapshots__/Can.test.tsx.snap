// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@pismo/bolt-can if the user has the permissions should display the components 1`] = `
<div>
  permitido
</div>
`;
